### PR TITLE
feat(javascore): implement IBCModule on xCall app

### PR DESCRIPTION
**Pull request by redlarva**
### Commit Message

```bash
type: commit message
```

see the [guidelines](https://github.com/icon-project/community/blob/main/guidelines/technical/software-development-guidelines.md#commit-messages) for commit messages.

### Changelog Entry

```bash
version: <log entry>
```

## Checklist:

- [x] I have performed a self-review of my own code
- [ ] I have documented my code in accordance with the [documentation guidelines](https://github.com/icon-project/community/blob/main/guidelines/technical/software-development-guidelines.md#documentation)
- [ ] My changes generate no new warnings
- [ ] I have added tests that prove my fix is effective or that my feature works
- [x] I have run the unit tests
- [ ] I only have one commit (if not, squash them into one commit).
- [ ] I have a descriptive commit message that adheres to the [commit message guidelines](https://github.com/icon-project/community/blob/main/guidelines/technical/software-development-guidelines.md#commit-messages)

> Please review the [CONTRIBUTING.md](/CONTRIBUTING.md) file for detailed contributing guidelines.

### CODE REVIEW

File: contracts/javascore/xcall/src/main/java/foundation/icon/btp/xcall/CallServiceImpl.java [hunk 1]
```diff
@@ -114,36 +101,38 @@ private void cleanupCallRequest(BigInteger sn) {
     public BigInteger sendCallMessage(String _to, byte[] _data, @Optional byte[] _rollback) {
         BigInteger sn = getNextSn();
         Address caller = Context.getCaller();
-        Context.require(caller.isContract() || _rollback == null, "RollbackNotPossible");
+        Context.require(caller.isContract(), "Only contract is allowed to send call message ");
```
Comment (AntonAndell) [inline contracts/javascore/xcall/src/main/java/foundation/icon/btp/xcall/CallServiceImpl.java:104]: This change should be reverted. Caller only need to be contracts if rollback is passed
https://github.com/icon-project/IIPs/issues/52#issuecomment-1384866064

File: contracts/javascore/xcall/src/main/java/foundation/icon/btp/xcall/CallServiceImpl.java [hunk 2]
```diff
@@ -114,36 +101,38 @@ private void cleanupCallRequest(BigInteger sn) {
...
-        hgt.setRevisionHeight(timeoutHeight.get());
-        hgt.setRevisionNumber(BigInteger.ZERO);
+        BigInteger timeoutHeight = BigInteger.valueOf(Context.getBlockHeight()).add(_timoutHeight);
+        hgt.setRevisionHeight(timeoutHeight);
+
         pct.setTimeoutHeight(hgt);
 
-        pct.setTimeoutTimestamp(BigInteger.ZERO);
+        //TODO set packet timeout
+        BigInteger timeout = BigInteger.valueOf(Context.getBlockTimestamp())
```
Comment (AntonAndell) [inline contracts/javascore/xcall/src/main/java/foundation/icon/btp/xcall/CallServiceImpl.java:131]: Keep this as 0 instead which in IBC means can't be timed out by timestamp. Only height

File: contracts/javascore/xcall/src/main/java/foundation/icon/btp/xcall/CallServiceImpl.java [hunk 3]
```diff
@@ -374,34 +390,31 @@ public void onChanCloseConfirm(String portId, String channelId) {
...
+        recvCount.set(newRecvCount);
 
+        Packet packet = Packet.decode(calldata);
+        byte[] _msg = packet.getData();
+        String _from = packet.getSourcePort() + "/" + packet.getSourceChannel();
+        BigInteger _sn = packet.getSequence();
 
-        return new byte[0];
-
+        handleReceivedPacket(_from, _from, _sn, _msg);
```
Comment (AntonAndell) [inline contracts/javascore/xcall/src/main/java/foundation/icon/btp/xcall/CallServiceImpl.java:404]: Ports and channel could be checked here to be correct

File: contracts/javascore/lib/src/main/java/ibc/icon/interfaces/IMock.java [hunk 4]
```diff
@@ -0,0 +1,8 @@
+package ibc.icon.interfaces;
+
+import foundation.icon.score.client.ScoreInterface;
+
+@ScoreInterface
+public interface IMock {
```
Comment (AntonAndell) [inline contracts/javascore/lib/src/main/java/ibc/icon/interfaces/IMock.java:6]: Is this used anywhere?

**Review by AntonAndell — CHANGES_REQUESTED** [review]
Looks like a good inital draft of XCall that we can start testing with.
Just some minor comments.

File: contracts/javascore/lib/src/main/java/ibc/icon/interfaces/IMock.java (hunk 4 above)
Comment (redlarva) [reply contracts/javascore/lib/src/main/java/ibc/icon/interfaces/IMock.java:6]: Yes it is used create empty MockContract for contract address.

File: contracts/javascore/xcall/src/main/java/foundation/icon/btp/xcall/CallServiceImpl.java (hunk 3 above)
Comment (redlarva) [reply contracts/javascore/xcall/src/main/java/foundation/icon/btp/xcall/CallServiceImpl.java]: done

File: contracts/javascore/xcall/src/main/java/foundation/icon/btp/xcall/CallServiceImpl.java (hunk 2 above)
Comment (redlarva) [reply contracts/javascore/xcall/src/main/java/foundation/icon/btp/xcall/CallServiceImpl.java]: done

File: contracts/javascore/xcall/src/main/java/foundation/icon/btp/xcall/CallServiceImpl.java [hunk 5]
```diff
@@ -114,36 +101,38 @@ private void cleanupCallRequest(BigInteger sn) {
...
         Context.require(_data.length <= MAX_DATA_SIZE, "MaxDataSizeExceeded");
         Context.require(_rollback == null || _rollback.length <= MAX_ROLLBACK_SIZE, "MaxRollbackSizeExceeded");
 
-        boolean needResponse = _rollback != null;
+        boolean needResponse = _rollback != null && _rollback.length > 0;
 
-        if (needResponse) {
-            CallRequest req = new CallRequest(caller, _to, _rollback);
-            requests.set(sn, req);
-        }
```
[the remark this answers is not in the document]
Comment (redlarva) [reply contracts/javascore/xcall/src/main/java/foundation/icon/btp/xcall/CallServiceImpl.java]: done

File: contracts/javascore/xcall/src/main/java/foundation/icon/btp/xcall/CallServiceImpl.java (hunk 1 above)
Comment (redlarva) [reply contracts/javascore/xcall/src/main/java/foundation/icon/btp/xcall/CallServiceImpl.java]: done

File: contracts/javascore/lib/src/main/java/ibc/icon/interfaces/IMock.java (hunk 4 above)
Comment (AntonAndell) [reply contracts/javascore/lib/src/main/java/ibc/icon/interfaces/IMock.java:6]: You can use sm.createScoreAccount for this,  MockContracts are only useful to mock and verify method calls.

**Review by AntonAndell — APPROVED** [review]
Looks good :+1:

**Pull request merged**